drizzlepac/haputils/product.py

```diff
@@ -127,6 +131,222 @@ def generate_metawcs(self):
+                        # fit geometry, but it may be downgraded depending upon the number of
+                        # sources available.
+                        if not fitgeom.strip():
+                            mosaic_fitgeom = alignment_pars[full_method_name]['fitgeom']
+                        else:
+                            mosaic_fitgeom = fitgeom
+
+                        # Get the index of the fitgeom in the list of all possible fitgeoms
+                        for index, mode in enumerate(mosaic_fitgeom_list):
+                            if mode[0].lower().strip() == mosaic_fitgeom:
+                                mosaic_fitgeom_index = index
+                                break
+
+                        # If there are not enough references sources for the specified fitgeom,
+                        # downgrade the fitgeom until a valid one is found.  Also, if the fit done
+                        # with the fitgeom was unsatisfactory, downgrade if possible and try again.
+                        # pdb.set_trace()
+                        while num_ref_sources < mosaic_fitgeom_list[mosaic_fitgeom_index][1]:
+                            log.warning("Not enough reference sources for alignment using catalog '{}' with fit method '{}' and fit geometry '{}'.".format(catalog_item, method_name, mosaic_fitgeom))
+                            mosaic_fitgeom_index = -1
+                            if mosaic_fitgeom_index < 0:
+                                log.warning("No further fit geometries to try. Proceeding to try another fit method.")
+                                break
+
+                        while mosaic_fitgeom_index > -1:
+
+                            # In case of a downgrade, make sure the fit geometry "name" is based upon the index
+                            mosaic_fitgeom = mosaic_fitgeom_list[mosaic_fitgeom_index][0]
+                            log.info("Proceeding to use the '{}' fit geometry.".format(mosaic_fitgeom))
+
+                            try:
+                                # Always perform the configuration to ensure there is no residual cruft
+                                # from any possible previous fit
+                                align_table.configure_fit()
+
+                                log.info("Trying '{}' wth method '{}' using fit geometry '{}'.".
+                                         format(catalog_item, method_name, mosaic_fitgeom))
+                                # pdb.set_trace()
+                                align_table.imglist = align_table.perform_fit(method_name, catalog_item, ref_catalog,
+                                                                              fitgeom=mosaic_fitgeom)
+
+                                align_table.select_fit(catalog_item, method_name)
+                                align_table.apply_fit(headerlet_filenames=headerlet_filenames,
+                                                      fit_label=fit_label)
```

Comment:
Selecting a fit at this point will always result in the first attempt being defined as the 'best' (final) solution, then subsequent calls with other methods and catalogs will return an Exception.  Also, since 'determine_fits_quality_mvm_interface' has not been run yet, the determination has not yet been made as to whether the fit in 'select_fit' is even good.  These lines can only be called  after 'determine_fit...' at line 286.

Reply:
Issues corrected.